js/dispatch.ts

```diff
@@ -5,35 +5,86 @@ import * as msg from "gen/msg_generated";
 import * as errors from "./errors";
 import * as util from "./util";
 import { maybePushTrace } from "./trace";
+import { promiseErrorExaminer } from "./promise_util";
 
 let nextCmdId = 0;
 const promiseTable = new Map<number, util.Resolvable<msg.Base>>();
 
-let fireTimers: () => void;
+let fireTimers: () => number;
+let nTasks = 0; // Number of async tasks pending.
+let delay = 0; // Cached return value of fireTimers.
 
-export function setFireTimersCallback(fn: () => void) {
+function eventLoopLog(): void {
+  util.log(`TICK delay ${delay} nTasks ${nTasks}`);
+}
+
+function idle(): boolean {
+  delay = fireTimers();
+  return delay < 0 && nTasks === 0;
+}
+
+export function eventLoop(): boolean {
+  for (;;) {
```

Comment:
I always liked `while (true)` better for infinite loops, feels more intentional to me, but...